### PR TITLE
[20262] Add non-throwing getters for socket info

### PR DESCRIPTION
**Pull request by jepemi**
## Description

In some instances, requesting server info after client connection caused exceptions (transport endpoint is not connected). This behavior was observed after PR #4258.
We now also use ASIO's non-throwing methods for this request, ensuring any errors are appropriately handled.

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- [x] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- N/A New feature has been added to the `versions.md` file (if applicable).
- N/A New feature has been documented/Current behavior is correctly described in the documentation. 
    
- [ ] Applicable backports have been included in the description.

## Reviewer Checklist
- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [ ] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: src/cpp/rtps/transport/TCPChannelResourceBasic.h [hunk 1]
```diff
@@ -65,9 +65,16 @@ class TCPChannelResourceBasic : public TCPChannelResource
             size_t size,
             asio::error_code& ec) override;
 
+    // Non-throwing asio calls
     asio::ip::tcp::endpoint remote_endpoint() const override;
     asio::ip::tcp::endpoint local_endpoint() const override;
 
+    // Throwing asio calls
```
Comment (MiguelCompany) [inline src/cpp/rtps/transport/TCPChannelResourceBasic.h:72]:
```suggestion
    // Throwing asio calls
    asio::ip::tcp::endpoint remote_endpoint() const override;
    asio::ip::tcp::endpoint local_endpoint() const override;

    // Non-throwing asio calls
```

File: src/cpp/rtps/transport/TCPTransportInterface.cpp [hunk 2]
```diff
@@ -898,7 +898,12 @@ void TCPTransportInterface::perform_listen_operation(
     if (rtcp_message_manager)
     {
         channel = channel_weak.lock();
-        endpoint_to_locator(channel->remote_endpoint(), remote_locator);
+        asio::error_code ec;
+        endpoint_to_locator(channel->remote_endpoint(ec), remote_locator);
+        if (ec)
+        {
+            remote_locator.kind = LOCATOR_KIND_INVALID;
+        }
```
Comment (MiguelCompany) [inline src/cpp/rtps/transport/TCPTransportInterface.cpp:906]: I would have a new method `Locator remote_endpoint_to_locator(channel)` that performs all these steps.

File: src/cpp/rtps/transport/TCPTransportInterface.cpp [hunk 3]
```diff
@@ -1178,6 +1183,10 @@ bool TCPTransportInterface::Receive(
                     }
                     else
                     {
+                        if (!IsLocatorValid(remote_locator))
+                        {
+                            endpoint_to_locator(channel->remote_endpoint(), remote_locator);
```
Comment (MiguelCompany) [inline src/cpp/rtps/transport/TCPTransportInterface.cpp:1188]: Also use the new method here ?

File: src/cpp/rtps/transport/TCPChannelResource.h [hunk 4]
```diff
@@ -134,6 +134,12 @@ class TCPChannelResource : public ChannelResource
 
     virtual asio::ip::tcp::endpoint local_endpoint() const = 0;
 
+    virtual asio::ip::tcp::endpoint remote_endpoint(
+            asio::error_code& ec) const = 0;
+
+    virtual asio::ip::tcp::endpoint local_endpoint(
+            asio::error_code& ec) const = 0;
+
```
Comment (MiguelCompany) [inline src/cpp/rtps/transport/TCPChannelResource.h:142]: Please document the new methods with Doxygen

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Comment by jepemi** [comment]
@richiprosima please test this

File: src/cpp/rtps/transport/TCPTransportInterface.cpp [hunk 5]
```diff
@@ -898,12 +911,8 @@ void TCPTransportInterface::perform_listen_operation(
...
     {
         channel = channel_weak.lock();
-        asio::error_code ec;
-        endpoint_to_locator(channel->remote_endpoint(ec), remote_locator);
-        if (ec)
-        {
-            remote_locator.kind = LOCATOR_KIND_INVALID;
-        }
+
+        remote_locator = remote_endpoint_to_locator(channel);
```
Comment (MiguelCompany) [inline src/cpp/rtps/transport/TCPTransportInterface.cpp:915]: We should perform this call inside the `if (channel)` below

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Review by MiguelCompany — APPROVED** [review]

**Comment by JesusPoderoso** [comment]
@mergifyio rebase

**Comment by JesusPoderoso** [comment]
@richiprosima please test this

**Pull request merged**